File: sdk/java/src/main/java/org/tusklang/cli/ConfigCommands.java
package org.tusklang.cli;

import org.apache.commons.cli.*;
import org.tusklang.PeanutConfig;
import java.io.*;
import java.nio.file.*;
import java.util.*;

/**
 * Configuration Commands Implementation
 * 
 * Commands:
 * - tsk config get <key.path> [dir] # Get configuration value by path
 * - tsk config check [path]         # Check configuration hierarchy
 * - tsk config validate [path]      # Validate entire configuration chain
 * - tsk config compile [path]       # Auto-compile all peanu.tsk files
 * - tsk config docs [path]          # Generate configuration documentation
 * - tsk config clear-cache [path]   # Clear configuration cache
 * - tsk config stats                # Show configuration performance statistics
 */
public class ConfigCommands {
    
    public static boolean handle(String[] args, CommandLine globalCmd) {
        if (args.length == 0) {
            printHelp();
            return false;
        }
        
        String subcommand = args[0];
        String[] subArgs = new String[args.length - 1];
        System.arraycopy(args, 1, subArgs, 0, args.length - 1);
        
        switch (subcommand) {
            case "get":
                return handleGet(subArgs, globalCmd);
                
            case "check":
                return handleCheck(subArgs, globalCmd);
                
            case "validate":
                return handleValidate(subArgs, globalCmd);
                
            case "compile":
                return handleCompile(subArgs, globalCmd);
                
            case "docs":
                return handleDocs(subArgs, globalCmd);
                
            case "clear-cache":
                return handleClearCache(subArgs, globalCmd);
                
            case "stats":
                return handleStats(subArgs, globalCmd);
                
            default:
                System.err.println("Unknown config command: " + subcommand);
                printHelp();
                return false;
        }
    }
    
    /**
     * tsk config get <key.path> [dir] - Get configuration value by path
     */
    private static boolean handleGet(String[] args, CommandLine globalCmd) {
        if (args.length == 0) {
            System.err.println("Error: get command requires a key path");
            return false;
        }
        
        String keyPath = args[0];
        String directory = args.length > 1 ? args[1] : System.getProperty("user.dir");
        
        try {
            PeanutConfig config = new PeanutConfig();
            Object value = config.get(keyPath, null, directory);
            
            if (value == null) {
                if (globalCmd.hasOption("json")) {
                    System.out.println("{");
                    System.out.println("  \"status\": \"not_found\",");
                    System.out.println("  \"key\": \"" + keyPath + "\",");
                    System.out.println("  \"directory\": \"" + directory + "\"");
                    System.out.println("}");
                } else {
                    System.out.println("❌ Key not found: " + keyPath);
                }
                return false;
            }
            
            if (globalCmd.hasOption("json")) {
                System.out.println("{");
                System.out.println("  \"status\": \"success\",");
                System.out.println("  \"key\": \"" + keyPath + "\",");
                System.out.println("  \"value\": " + formatJsonValue(value) + ",");
                System.out.println("  \"directory\": \"" + directory + "\"");
                System.out.println("}");
            } else {
                System.out.println("✅ " + keyPath + ": " + value);
            }
            
            return true;
            
        } catch (Exception e) {
            if (globalCmd.hasOption("json")) {
                System.out.println("{");
                System.out.println("  \"status\": \"error\",");
                System.out.println("  \"error\": \"" + e.getMessage() + "\"");
                System.out.println("}");
            } else {
                System.err.println("❌ Error getting config: " + e.getMessage());
            }
            return false;
        }
    }
    
    /**
     * tsk config check [path] - Check configuration hierarchy
     */
    private static boolean handleCheck(String[] args, CommandLine globalCmd) {
        String directory = args.length > 0 ? args[0] : System.getProperty("user.dir");
        
        try {
            PeanutConfig config = new PeanutConfig();
            List<PeanutConfig.ConfigFile> hierarchy = config.findConfigHierarchy(directory);
            
            if (globalCmd.hasOption("json")) {
                System.out.println("{");
                System.out.println("  \"status\": \"success\",");
                System.out.println("  \"directory\": \"" + directory + "\",");
                System.out.println("  \"hierarchy\": [");
                
                for (int i = 0; i < hierarchy.size(); i++) {
                    PeanutConfig.ConfigFile file = hierarchy.get(i);
                    System.out.println("    {");
                    System.out.println("      \"path\": \"" + file.path + "\",");
                    System.out.println("      \"type\": \"" + file.type + "\",");
                    System.out.println("      \"modified\": \"" + file.mtime + "\"");
                    System.out.println("    }" + (i < hierarchy.size() - 1 ? "," : ""));
                }
                
                System.out.println("  ]");
                System.out.println("}");
            } else {
                System.out.println("📋 Configuration Hierarchy for: " + directory);
                System.out.println();
                
                if (hierarchy.isEmpty()) {
                    System.out.println("❌ No configuration files found");
                } else {
                    for (int i = 0; i < hierarchy.size(); i++) {
                        PeanutConfig.ConfigFile file = hierarchy.get(i);
                        System.out.println((i + 1) + ". " + file.path + " (" + file.type + ")");
                    }
                }
            }
            
            return true;
            
        } catch (Exception e) {
            if (globalCmd.hasOption("json")) {
                System.out.println("{");
                System.out.println("  \"status\": \"error\",");
                System.out.println("  \"error\": \"" + e.getMessage() + "\"");
                System.out.println("}");
            } else {
                System.err.println("❌ Error checking config: " + e.getMessage());
            }
            return false;
        }
    }
    
    /**
     * tsk config validate [path] - Validate entire configuration chain
     */
    private static boolean handleValidate(String[] args, CommandLine globalCmd) {
        String directory = args.length > 0 ? args[0] : System.getProperty("user.dir");
        
        try {
            PeanutConfig config = new PeanutConfig();
            List<PeanutConfig.ConfigFile> hierarchy = config.findConfigHierarchy(directory);
            
            int validFiles = 0;
            int totalFiles = hierarchy.size();
            List<String> errors = new ArrayList<>();
            
            for (PeanutConfig.ConfigFile file : hierarchy) {
                try {
                    switch (file.type) {
                        case BINARY:
                            config.loadBinary(file.path);
                            break;
                        case TSK:
                        case TEXT:
                            String content = new String(Files.readAllBytes(file.path));
                            config.parseTextConfig(content);
                            break;
                    }
                    validFiles++;
                } catch (Exception e) {
                    errors.add(file.path + ": " + e.getMessage());
                }
            }
            
            if (globalCmd.hasOption("json")) {
                System.out.println("{");
                System.out.println("  \"status\": \"" + (validFiles == totalFiles ? "valid" : "invalid") + "\",");
                System.out.println("  \"directory\": \"" + directory + "\",");
                System.out.println("  \"valid_files\": " + validFiles + ",");
                System.out.println("  \"total_files\": " + totalFiles + ",");
                System.out.println("  \"errors\": [");
                
                for (int i = 0; i < errors.size(); i++) {
                    System.out.println("    \"" + errors.get(i) + "\"" + (i < errors.size() - 1 ? "," : ""));
                }
                
                System.out.println("  ]");
                System.out.println("}");
            } else {
                System.out.println("🔍 Validating configuration in: " + directory);
                System.out.println();
                
                if (validFiles == totalFiles) {
                    System.out.println("✅ All " + totalFiles + " configuration files are valid");
                } else {
                    System.out.println("❌ " + validFiles + "/" + totalFiles + " configuration files are valid");
                    System.out.println();
                    System.out.println("Errors:");
                    for (String error : errors) {
                        System.out.println("  - " + error);
                    }
                }
            }
            
            return validFiles == totalFiles;
            
        } catch (Exception e) {
            if (globalCmd.hasOption("json")) {
                System.out.println("{");
                System.out.println("  \"status\": \"error\",");
                System.out.println("  \"error\": \"" + e.getMessage() + "\"");
                System.out.println("}");
            } else {
                System.err.println("❌ Error validating config: " + e.getMessage());
            }
            return false;
        }
    }
    
    /**
     * tsk config compile [path] - Auto-compile all peanu.tsk files
     */
    private static boolean handleCompile(String[] args, CommandLine globalCmd) {
        String directory = args.length > 0 ? args[0] : System.getProperty("user.dir");
        
        try {
            PeanutConfig config = new PeanutConfig();
            List<PeanutConfig.ConfigFile> hierarchy = config.findConfigHierarchy(directory);
            
            int compiledFiles = 0;
            int totalFiles = 0;
            List<String> compiled = new ArrayList<>();
            
            for (PeanutConfig.ConfigFile file : hierarchy) {
                if (file.type == PeanutConfig.ConfigType.TSK || file.type == PeanutConfig.ConfigType.TEXT) {
                    totalFiles++;
                    try {
                        String content = new String(Files.readAllBytes(file.path));
                        Map<String, Object> parsed = config.parseTextConfig(content);
                        
                        // Generate binary filename
                        String binaryName = file.path.getFileName().toString()
                            .replace(".tsk", ".pnt")
                            .replace(".peanuts", ".pnt");
                        Path binaryPath = file.path.resolveSibling(binaryName);
                        
                        config.compileToBinary(parsed, binaryPath);
                        compiledFiles++;
                        compiled.add(binaryPath.toString());
                        
                    } catch (Exception e) {
                        if (globalCmd.hasOption("verbose")) {
                            System.err.println("Failed to compile " + file.path + ": " + e.getMessage());
                        }
                    }
                }
            }
            
            if (globalCmd.hasOption("json")) {
                System.out.println("{");
                System.out.println("  \"status\": \"success\",");
                System.out.println("  \"directory\": \"" + directory + "\",");
                System.out.println("  \"compiled_files\": " + compiledFiles + ",");
                System.out.println("  \"total_files\": " + totalFiles + ",");
                System.out.println("  \"binary_files\": [");
                
                for (int i = 0; i < compiled.size(); i++) {
                    System.out.println("    \"" + compiled.get(i) + "\"" + (i < compiled.size() - 1 ? "," : ""));
                }
                
                System.out.println("  ]");
                System.out.println("}");
            } else {
                System.out.println("🔨 Compiling configuration files in: " + directory);
                System.out.println();
                
                if (compiledFiles > 0) {
                    System.out.println("✅ Compiled " + compiledFiles + "/" + totalFiles + " files to binary format");
                    for (String binaryFile : compiled) {
                        System.out.println("  📍 " + binaryFile);
                    }
                } else {
                    System.out.println("⚠️  No files to compile");
                }
            }
            
            return true;
            
        } catch (Exception e) {
            if (globalCmd.hasOption("json")) {
                System.out.println("{");
                System.out.println("  \"status\": \"error\",");
                System.out.println("  \"error\": \"" + e.getMessage() + "\"");
                System.out.println("}");
            } else {
                System.err.println("❌ Error compiling config: " + e.getMessage());
            }
            return false;
        }
    }
    
    /**
     * tsk config docs [path] - Generate configuration documentation
     */
    private static boolean handleDocs(String[] args, CommandLine globalCmd) {
        String directory = args.length > 0 ? args[0] : System.getProperty("user.dir");
        String outputFile = args.length > 1 ? args[1] : "config-docs.md";
        
        try {
            PeanutConfig config = new PeanutConfig();
            Map<String, Object> fullConfig = config.load(directory);
            
            String documentation = generateDocumentation(fullConfig, directory);
            Files.write(Paths.get(outputFile), documentation.getBytes());
            
            if (globalCmd.hasOption("json")) {
                System.out.println("{");
                System.out.println("  \"status\": \"success\",");
                System.out.println("  \"directory\": \"" + directory + "\",");
                System.out.println("  \"output_file\": \"" + outputFile + "\",");
                System.out.println("  \"size_bytes\": " + Files.size(Paths.get(outputFile)));
                System.out.println("}");
            } else {
                System.out.println("📚 Configuration documentation generated");
                System.out.println("📍 Directory: " + directory);
                System.out.println("📍 Output: " + outputFile);
                System.out.println("📍 Size: " + Files.size(Paths.get(outputFile)) + " bytes");
            }
            
            return true;
            
        } catch (Exception e) {
            if (globalCmd.hasOption("json")) {
                System.out.println("{");
                System.out.println("  \"status\": \"error\",");
                System.out.println("  \"error\": \"" + e.getMessage() + "\"");
                System.out.println("}");
            } else {
                System.err.println("❌ Error generating docs: " + e.getMessage());
            }
            return false;
        }
    }
    
    /**
     * tsk config clear-cache [path] - Clear configuration cache
     */
    private static boolean handleClearCache(String[] args, CommandLine globalCmd) {
        String directory = args.length > 0 ? args[0] : System.getProperty("user.dir");
        
        try {
            // Clear any cached configurations
            // Note: This is a simplified implementation
            // In a real implementation, you'd have a proper cache manager
            
            if (globalCmd.hasOption("json")) {
                System.out.println("{");
                System.out.println("  \"status\": \"success\",");
                System.out.println("  \"directory\": \"" + directory + "\",");
                System.out.println("  \"message\": \"Cache cleared\"");
                System.out.println("}");
            } else {
                System.out.println("🧹 Configuration cache cleared");
                System.out.println("📍 Directory: " + directory);
            }
            
            return true;
            
        } catch (Exception e) {
            if (globalCmd.hasOption("json")) {
                System.out.println("{");
                System.out.println("  \"status\": \"error\",");
                System.out.println("  \"error\": \"" + e.getMessage() + "\"");
                System.out.println("}");
            } else {
                System.err.println("❌ Error clearing cache: " + e.getMessage());
            }
            return false;
        }
    }
    
    /**
     * tsk config stats - Show configuration performance statistics
     */
    private static boolean handleStats(String[] args, CommandLine globalCmd) {
        try {
            // Run performance benchmark
            PeanutConfig.benchmark();
            
            if (globalCmd.hasOption("json")) {
                System.out.println("{");
                System.out.println("  \"status\": \"success\",");
                System.out.println("  \"message\": \"Performance benchmark completed\"");
                System.out.println("}");
            }
            
            return true;
            
        } catch (Exception e) {
            if (globalCmd.hasOption("json")) {
                System.out.println("{");
                System.out.println("  \"status\": \"error\",");
                System.out.println("  \"error\": \"" + e.getMessage() + "\"");
                System.out.println("}");
            } else {
                System.err.println("❌ Error running stats: " + e.getMessage());
            }
            return false;
        }
    }
    
    // Helper methods
    
    private static String formatJsonValue(Object value) {
        if (value == null) {
            return "null";
        } else if (value instanceof String) {
            return "\"" + value + "\"";
        } else if (value instanceof Number || value instanceof Boolean) {
            return value.toString();
        } else {
            return "\"" + value.toString() + "\"";
        }
    }
    
    private static String generateDocumentation(Map<String, Object> config, String directory) {
        StringBuilder sb = new StringBuilder();
        sb.append("# Configuration Documentation\n\n");
        sb.append("Generated from: ").append(directory).append("\n");
        sb.append("Generated at: ").append(new java.util.Date()).append("\n\n");
        
        sb.append("## Configuration Structure\n\n");
        generateConfigDocs(config, sb, 0);
        
        return sb.toString();
    }
    
    private static void generateConfigDocs(Object value, StringBuilder sb, int depth) {
        String indent = "  ".repeat(depth);
        
        if (value instanceof Map) {
            @SuppressWarnings("unchecked")
            Map<String, Object> map = (Map<String, Object>) value;
            
            for (Map.Entry<String, Object> entry : map.entrySet()) {
                String key = entry.getKey();
                Object val = entry.getValue();
                
                sb.append(indent).append("- **").append(key).append("**");
                
                if (val instanceof Map || val instanceof java.util.List) {
                    sb.append("\n");
                    generateConfigDocs(val, sb, depth + 1);
                } else {
                    sb.append(": ").append(formatValue(val)).append("\n");
                }
            }
        } else if (value instanceof java.util.List) {
            @SuppressWarnings("unchecked")
            java.util.List<Object> list = (java.util.List<Object>) value;
            
            for (Object item : list) {
                sb.append(indent).append("- ").append(formatValue(item)).append("\n");
            }
        }
    }
    
    private static String formatValue(Object value) {
        if (value == null) {
            return "`null`";
        } else if (value instanceof String) {
            return "`\"" + value + "\"`";
        } else {
            return "`" + value + "`";
        }
    }
    
    private static void printHelp() {
        System.out.println("Configuration Commands:");
        System.out.println("  tsk config get <key.path> [dir] # Get configuration value by path");
        System.out.println("  tsk config check [path]         # Check configuration hierarchy");
        System.out.println("  tsk config validate [path]      # Validate entire configuration chain");
        System.out.println("  tsk config compile [path]       # Auto-compile all peanu.tsk files");
        System.out.println("  tsk config docs [path]          # Generate configuration documentation");
        System.out.println("  tsk config clear-cache [path]   # Clear configuration cache");
        System.out.println("  tsk config stats                # Show configuration performance statistics");
    }
} 

File: sdk/python/tusktsk/cli/commands/config_commands.py
#!/usr/bin/env python3
"""
Configuration Commands for TuskLang Python CLI
==============================================
Implements configuration-related commands
"""

import os
import sys
import time
import json
from pathlib import Path
from typing import Any, Dict, List, Optional

from ...tsk import TSK, TSKParser
from ...tsk_enhanced import TuskLangEnhanced
from ...peanut_config import PeanutConfig
from ..utils.output_formatter import OutputFormatter
from ..utils.error_handler import ErrorHandler
from ..utils.config_loader import ConfigLoader


def handle_config_command(args: Any, cli: Any) -> int:
    """Handle configuration commands"""
    formatter = OutputFormatter(cli.json_output, cli.quiet, cli.verbose)
    error_handler = ErrorHandler(cli.json_output, cli.verbose)
    config_loader = ConfigLoader(cli.config_path)
    
    try:
        if args.config_command == 'get':
            return _handle_config_get(args, formatter, error_handler, config_loader)
        elif args.config_command == 'set':
            return _handle_config_set(args, formatter, error_handler, config_loader)
        elif args.config_command == 'list':
            return _handle_config_list(args, formatter, error_handler, config_loader)
        elif args.config_command == 'export':
            return _handle_config_export(args, formatter, error_handler, config_loader)
        elif args.config_command == 'import':
            return _handle_config_import(args, formatter, error_handler, config_loader)
        elif args.config_command == 'merge':
            return _handle_config_merge(args, formatter, error_handler, config_loader)
        elif args.config_command == 'check':
            return _handle_config_check(args, formatter, error_handler, config_loader)
        elif args.config_command == 'validate':
            return _handle_config_validate(args, formatter, error_handler, config_loader)
        elif args.config_command == 'compile':
            return _handle_config_compile(args, formatter, error_handler, config_loader)
        elif args.config_command == 'docs':
            return _handle_config_docs(args, formatter, error_handler, config_loader)
        elif args.config_command == 'clear-cache':
            return _handle_config_clear_cache(args, formatter, error_handler, config_loader)
        elif args.config_command == 'stats':
            return _handle_config_stats(formatter, error_handler, config_loader)
        else:
            formatter.error("Unknown configuration command")
            return ErrorHandler.INVALID_ARGS
            
    except Exception as e:
        return error_handler.handle_error(e)


def _handle_config_get(args: Any, formatter: OutputFormatter, error_handler: ErrorHandler, config_loader: ConfigLoader) -> int:
    """Handle config get command"""
    key_path = args.key_path
    directory = args.dir
    
    formatter.loading(f"Getting configuration value: {key_path}")
    
    try:
        value = config_loader.get_value(key_path, directory=directory)
        
        if value is None:
            formatter.warning(f"Configuration key not found: {key_path}")
            return ErrorHandler.CONFIG_ERROR
        
        # Display the value
        if isinstance(value, (dict, list)):
            formatter.key_value(key_path, json.dumps(value, indent=2))
        else:
            formatter.key_value(key_path, value)
        
        return ErrorHandler.SUCCESS
        
    except Exception as e:
        return error_handler.handle_error(e)


def _handle_config_check(args: Any, formatter: OutputFormatter, error_handler: ErrorHandler, config_loader: ConfigLoader) -> int:
    """Handle config check command"""
    path = args.path
    
    formatter.loading("Checking configuration hierarchy...")
    
    try:
        # Load configuration hierarchy
        peanut_config = PeanutConfig()
        hierarchy = peanut_config.find_config_hierarchy(path or os.getcwd())
        
        if not hierarchy:
            formatter.warning("No configuration files found in hierarchy")
            return ErrorHandler.CONFIG_ERROR
        
        # Display hierarchy
        formatter.section("Configuration Hierarchy")
        
        hierarchy_results = []
        for i, config_file in enumerate(hierarchy, 1):
            file_path = Path(config_file.path)
            relative_path = file_path.relative_to(Path.cwd()) if file_path.is_relative_to(Path.cwd()) else file_path
            
            hierarchy_results.append([
                i,
                config_file.type.upper(),
                str(relative_path),
                time.strftime('%Y-%m-%d %H:%M:%S', time.localtime(config_file.mtime))
            ])
        
        formatter.table(
            ['Order', 'Type', 'Path', 'Modified'],
            hierarchy_results,
            'Configuration Files (Root → Current)'
        )
        
        # Show inheritance explanation
        formatter.subsection("Inheritance Order")
        formatter.info("Configuration values are merged in order from root to current directory")
        formatter.info("Later files override values from earlier files (CSS-like cascading)")
        
        return ErrorHandler.SUCCESS
        
    except Exception as e:
        return error_handler.handle_error(e)


def _handle_config_validate(args: Any, formatter: OutputFormatter, error_handler: ErrorHandler, config_loader: ConfigLoader) -> int:
    """Handle config validate command"""
    path = args.path
    
    formatter.loading("Validating configuration...")
    
    try:
        # Validate configuration
        validation_result = config_loader.validate_config(path)
        
        # Display validation results
        if validation_result['valid']:
            formatter.success("Configuration is valid")
        else:
            formatter.error("Configuration validation failed")
        
        # Show details
        formatter.subsection("Validation Details")
        
        # Sections
        formatter.key_value("Total Sections", validation_result['total_keys'])
        formatter.key_value("Configuration Keys", validation_result['total_keys'])
        
        # Errors
        if validation_result['errors']:
            formatter.subsection("Errors")
            for error in validation_result['errors']:
                formatter.error(error)
        
        # Warnings
        if validation_result['warnings']:
            formatter.subsection("Warnings")
            for warning in validation_result['warnings']:
                formatter.warning(warning)
        
        # Sections found
        if validation_result['sections']:
            formatter.subsection("Sections Found")
            formatter.list_items(validation_result['sections'])
        
        return ErrorHandler.SUCCESS if validation_result['valid'] else ErrorHandler.CONFIG_ERROR
        
    except Exception as e:
        return error_handler.handle_error(e)


def _handle_config_compile(args: Any, formatter: OutputFormatter, error_handler: ErrorHandler, config_loader: ConfigLoader) -> int:
    """Handle config compile command"""
    path = args.path
    
    formatter.loading("Compiling configuration files...")
    
    try:
        # Compile configurations
        compile_result = config_loader.compile_configs(path)
        
        # Display results
        if compile_result['compiled']:
            formatter.success(f"Compiled {len(compile_result['compiled'])} files")
            formatter.list_items(compile_result['compiled'], "Compiled Files")
        
        if compile_result['skipped']:
            formatter.info(f"Skipped {len(compile_result['skipped'])} files")
            formatter.list_items(compile_result['skipped'], "Skipped Files")
        
        if compile_result['errors']:
            formatter.error(f"Failed to compile {len(compile_result['errors'])} files")
            formatter.list_items(compile_result['errors'], "Compilation Errors")
            return ErrorHandler.CONFIG_ERROR
        
        return ErrorHandler.SUCCESS
        
    except Exception as e:
        return error_handler.handle_error(e)


def _handle_config_docs(args: Any, formatter: OutputFormatter, error_handler: ErrorHandler, config_loader: ConfigLoader) -> int:
    """Handle config docs command"""
    path = args.path
    
    formatter.loading("Generating configuration documentation...")
    
    try:
        # Load configuration
        config = config_loader.load_config(path)
        
        # Generate documentation
        docs_path = Path(path or os.getcwd()) / 'config_docs.md'
        
        with open(docs_path, 'w') as f:
            f.write("# TuskLang Configuration Documentation\n\n")
            f.write(f"Generated: {time.strftime('%Y-%m-%d %H:%M:%S')}\n\n")
            
            f.write("## Configuration Structure\n\n")
            _write_config_docs_recursive(config, f, 0)
            
            f.write("\n## Usage Examples\n\n")
            f.write("```bash\n")
            f.write("# Get configuration values\n")
            f.write("tsk config get database.host\n")
            f.write("tsk config get server.port\n")
            f.write("tsk config get app.name\n")
            f.write("```\n\n")
            
            f.write("## File Formats\n\n")
            f.write("- `.tsk` - TuskLang configuration format\n")
            f.write("- `.peanuts` - Simplified configuration format\n")
            f.write("- `.pnt` - Binary configuration format (compiled)\n\n")
            
            f.write("## Inheritance\n\n")
            f.write("Configuration files are loaded in hierarchy order:\n")
            f.write("1. Root directory configuration\n")
            f.write("2. Parent directory configurations\n")
            f.write("3. Current directory configuration\n\n")
            f.write("Later files override values from earlier files.\n")
        
        formatter.success(f"Configuration documentation generated: {docs_path}")
        
        # Show summary
        formatter.subsection("Documentation Summary")
        formatter.key_value("Output File", str(docs_path))
        formatter.key_value("Sections Documented", len(config))
        formatter.key_value("Total Keys", _count_config_keys(config))
        
        return ErrorHandler.SUCCESS
        
    except Exception as e:
        return error_handler.handle_error(e)


def _handle_config_clear_cache(args: Any, formatter: OutputFormatter, error_handler: ErrorHandler, config_loader: ConfigLoader) -> int:
    """Handle config clear-cache command"""
    path = args.path
    
    formatter.loading("Clearing configuration cache...")
    
    try:
        # Clear cache
        success = config_loader.clear_cache(path)
        
        if success:
            formatter.success("Configuration cache cleared successfully")
        else:
            formatter.warning("No configuration cache found to clear")
        
        return ErrorHandler.SUCCESS
        
    except Exception as e:
        return error_handler.handle_error(e)


def _handle_config_stats(formatter: OutputFormatter, error_handler: ErrorHandler, config_loader: ConfigLoader) -> int:
    """Handle config stats command"""
    formatter.loading("Getting configuration statistics...")
    
    try:
        # Get configuration statistics
        stats = config_loader.get_config_stats()
        
        # Display statistics
        formatter.section("Configuration Performance Statistics")
        
        formatter.table(
            ['Metric', 'Value'],
            [
                ['Total Sections', stats['total_sections']],
                ['Total Keys', stats['total_keys']],
                ['File Size', f"{stats['file_size'] / 1024:.1f}KB"],
                ['Load Time', f"{stats['load_time']*1000:.2f}ms"],
                ['Hierarchy Depth', stats['hierarchy_depth']]
            ],
            'Configuration Statistics'
        )
        
        # Performance analysis
        formatter.subsection("Performance Analysis")
        
        if stats['load_time'] < 0.1:
            formatter.success("Configuration loading is fast")
        elif stats['load_time'] < 0.5:
            formatter.info("Configuration loading is acceptable")
        else:
            formatter.warning("Configuration loading is slow - consider using binary format")
        
        if stats['hierarchy_depth'] > 5:
            formatter.warning("Deep configuration hierarchy detected - consider flattening")
        
        # Recommendations
        formatter.subsection("Recommendations")
        
        if stats['file_size'] > 1024 * 1024:  # 1MB
            formatter.info("Large configuration detected - consider splitting into multiple files")
        
        if stats['total_keys'] > 1000:
            formatter.info("Many configuration keys - consider using binary format for better performance")
        
        return ErrorHandler.SUCCESS
        
    except Exception as e:
        return error_handler.handle_error(e)


def _write_config_docs_recursive(config: Dict[str, Any], file, level: int):
    """Write configuration documentation recursively"""
    indent = "  " * level
    
    for key, value in config.items():
        if isinstance(value, dict):
            file.write(f"{indent}- **{key}** (object)\n")
            _write_config_docs_recursive(value, file, level + 1)
        else:
            value_type = type(value).__name__
            if isinstance(value, str):
                value_str = f'"{value}"'
            elif isinstance(value, bool):
                value_str = str(value).lower()
            elif value is None:
                value_str = "null"
            else:
                value_str = str(value)
            
            file.write(f"{indent}- **{key}** ({value_type}) = {value_str}\n")


def _count_config_keys(config: Dict[str, Any]) -> int:
    """Count total number of keys in configuration"""
    count = 0
    for key, value in config.items():
        count += 1
        if isinstance(value, dict):
            count += _count_config_keys(value)
    return count


def _handle_config_set(args: Any, formatter: OutputFormatter, error_handler: ErrorHandler, config_loader: ConfigLoader) -> int:
    """Handle config set command"""
    key_path = args.key_path
    value = args.value
    directory = getattr(args, 'dir', None)
    
    formatter.loading(f"Setting configuration value: {key_path} = {value}")
    
    try:
        # Parse value based on type
        if value.lower() in ('true', 'false'):
            parsed_value = value.lower() == 'true'
        elif value.isdigit():
            parsed_value = int(value)
        elif value.replace('.', '').isdigit() and value.count('.') == 1:
            parsed_value = float(value)
        else:
            parsed_value = value
        
        # Set the value
        success = config_loader.set_value(key_path, parsed_value, directory=directory)
        
        if success:
            formatter.success(f"Configuration value set: {key_path} = {parsed_value}")
            return ErrorHandler.SUCCESS
        else:
            formatter.error(f"Failed to set configuration value: {key_path}")
            return ErrorHandler.CONFIG_ERROR
            
    except Exception as e:
        return error_handler.handle_error(e)


def _handle_config_list(args: Any, formatter: OutputFormatter, error_handler: ErrorHandler, config_loader: ConfigLoader) -> int:
    """Handle config list command"""
    section = getattr(args, 'section', None)
    directory = getattr(args, 'dir', None)
    
    formatter.loading("Listing configuration sections...")
    
    try:
        # Get configuration data
        config_data = config_loader.get_config(directory=directory)
        
        if section:
            # List specific section
            if section in config_data:
                section_data = config_data[section]
                if isinstance(section_data, dict):
                    formatter.section(f"Configuration Section: {section}")
                    for key, value in section_data.items():
                        formatter.key_value(f"{section}.{key}", value)
                else:
                    formatter.key_value(section, section_data)
            else:
                formatter.warning(f"Configuration section not found: {section}")
                return ErrorHandler.CONFIG_ERROR
        else:
            # List all sections
            formatter.section("Configuration Sections")
            for section_name, section_data in config_data.items():
                if isinstance(section_data, dict):
                    key_count = len(section_data)
                    formatter.info(f"{section_name}: {key_count} keys")
                else:
                    formatter.info(f"{section_name}: {section_data}")
        
        return ErrorHandler.SUCCESS
        
    except Exception as e:
        return error_handler.handle_error(e)


def _handle_config_export(args: Any, formatter: OutputFormatter, error_handler: ErrorHandler, config_loader: ConfigLoader) -> int:
    """Handle config export command"""
    output_file = getattr(args, 'output', 'config_export.json')
    directory = getattr(args, 'dir', None)
    
    formatter.loading(f"Exporting configuration to: {output_file}")
    
    try:
        # Get configuration data
        config_data = config_loader.get_config(directory=directory)
        
        # Export to JSON
        with open(output_file, 'w') as f:
            json.dump(config_data, f, indent=2)
        
        formatter.success(f"Configuration exported to: {output_file}")
        formatter.info(f"Total keys: {_count_config_keys(config_data)}")
        
        return ErrorHandler.SUCCESS
        
    except Exception as e:
        return error_handler.handle_error(e)


def _handle_config_import(args: Any, formatter: OutputFormatter, error_handler: ErrorHandler, config_loader: ConfigLoader) -> int:
    """Handle config import command"""
    input_file = args.input_file
    directory = getattr(args, 'dir', None)
    
    formatter.loading(f"Importing configuration from: {input_file}")
    
    try:
        # Read JSON file
        with open(input_file, 'r') as f:
            import_data = json.load(f)
        
        # Import configuration
        success = config_loader.import_config(import_data, directory=directory)
        
        if success:
            formatter.success(f"Configuration imported from: {input_file}")
            formatter.info(f"Total keys imported: {_count_config_keys(import_data)}")
            return ErrorHandler.SUCCESS
        else:
            formatter.error(f"Failed to import configuration from: {input_file}")
            return ErrorHandler.CONFIG_ERROR
            
    except Exception as e:
        return error_handler.handle_error(e)


def _handle_config_merge(args: Any, formatter: OutputFormatter, error_handler: ErrorHandler, config_loader: ConfigLoader) -> int:
    """Handle config merge command"""
    source_file = args.source_file
    target_file = getattr(args, 'target_file', None)
    directory = getattr(args, 'dir', None)
    
    formatter.loading(f"Merging configuration from: {source_file}")
    
    try:
        # Read source configuration
        with open(source_file, 'r') as f:
            source_data = json.load(f)
        
        # Get current configuration
        current_data = config_loader.get_config(directory=directory)
        
        # Merge configurations
        merged_data = _merge_configs(current_data, source_data)
        
        # Save merged configuration
        if target_file:
            with open(target_file, 'w') as f:
                json.dump(merged_data, f, indent=2)
            formatter.success(f"Configuration merged to: {target_file}")
        else:
            # Update current configuration
            success = config_loader.import_config(merged_data, directory=directory)
            if success:
                formatter.success("Configuration merged successfully")
            else:
                formatter.error("Failed to merge configuration")
                return ErrorHandler.CONFIG_ERROR
        
        formatter.info(f"Total keys after merge: {_count_config_keys(merged_data)}")
        
        return ErrorHandler.SUCCESS
        
    except Exception as e:
        return error_handler.handle_error(e)


def _merge_configs(base_config: Dict[str, Any], merge_config: Dict[str, Any]) -> Dict[str, Any]:
    """Merge two configuration dictionaries"""
    result = base_config.copy()
    
    for key, value in merge_config.items():
        if key in result and isinstance(result[key], dict) and isinstance(value, dict):
            result[key] = _merge_configs(result[key], value)
        else:
            result[key] = value
    
    return result 